signer/tests/integration/block_observer.rs

```diff
@@ -906,3 +851,324 @@ async fn next_headers_to_process_ignores_known_headers() {
+    test_data.write_to(&db).await;
+
+    // We always need the chain tip.
+    let chain_tip = db.get_bitcoin_canonical_chain_tip().await.unwrap().unwrap();
+
+    // We have no rows in the DKG shares table and no rotate-keys
+    // transactions, so there should be no aggregate key, since that only
+    // happens after DKG, but we should always know the current signer set.
+    let (maybe_aggregate_key, signer_set) = get_signer_set_and_aggregate_key(&ctx, chain_tip)
+        .await
+        .unwrap();
+    assert!(maybe_aggregate_key.is_none());
+    assert!(!signer_set.is_empty());
+
+    // Alright, lets write some DKG shares into the database. When we do
+    // that the signer set should be considered whatever the signer set is
+    // from our DKG shares.
+    let shares: EncryptedDkgShares = Faker.fake_with_rng(&mut rng);
+    db.write_encrypted_dkg_shares(&shares).await.unwrap();
+
+    let (aggregate_key, signer_set) = get_signer_set_and_aggregate_key(&ctx, chain_tip)
+        .await
+        .unwrap();
+
+    let shares_signer_set: BTreeSet<PublicKey> =
+        shares.signer_set_public_keys.iter().copied().collect();
+
+    assert_eq!(shares.aggregate_key, aggregate_key.unwrap());
+    assert_eq!(shares_signer_set, signer_set);
+
+    // Okay not we write a rotate-keys transaction into the database. To do
+    // that we need the stacks chain tip, and a something in 3 different
+    // tables...
+    let stacks_chain_tip = db.get_stacks_chain_tip(&chain_tip).await.unwrap().unwrap();
+
+    let rotate_keys: RotateKeysTransaction = Faker.fake_with_rng(&mut rng);
+    let transaction = model::Transaction {
+        txid: rotate_keys.txid.into_bytes(),
+        tx: Vec::new(),
+        tx_type: model::TransactionType::RotateKeys,
+        block_hash: stacks_chain_tip.block_hash.into_bytes(),
+    };
+    let tx = model::StacksTransaction {
+        txid: rotate_keys.txid,
+        block_hash: stacks_chain_tip.block_hash,
+    };
+
+    db.write_transaction(&transaction).await.unwrap();
+    db.write_stacks_transaction(&tx).await.unwrap();
+    db.write_rotate_keys_transaction(&rotate_keys)
+        .await
+        .unwrap();
+
+    // Alright, now that we have a rotate-keys transaction, we can check if
+    // it is preferred over the DKG shares table.
+    let (aggregate_key, signer_set) = get_signer_set_and_aggregate_key(&ctx, chain_tip)
+        .await
+        .unwrap();
+
+    let rotate_keys_signer_set: BTreeSet<PublicKey> =
+        rotate_keys.signer_set.iter().copied().collect();
+
+    assert_eq!(rotate_keys.aggregate_key, aggregate_key.unwrap());
+    assert_eq!(rotate_keys_signer_set, signer_set);
+
+    testing::storage::drop_db(db).await;
+}
+
+/// This test checks that the signer state is updated with the latest the
+/// sbtc limits, current signer set, and current aggregate key after the
+/// block observer processes a bitcoin block.
+#[tokio::test]
+async fn block_observer_updates_state_after_observing_bitcoin_block() {
```

Comment:
Not suggesting to change it, but man it feels like a lot of work for testing this 🤔 When we get some time we should look into how we can break these kinds of tests into smaller units...